### ISSUE TITLE
Slow Code Completion for Function Names When Analyzing Overridable Functions in Classes

### ISSUE THREAD

**Issue by Aobanana-chan** [microsoft/pylance-release#5268]
<h3> Environment data </h3>

-   Language Server version: 2023.12.1
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda):  3.11.4
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: off

<h3> Code Snippet </h3>

```python
import libcst as cst
import libcst.matchers as m


class Transformer(m.MatcherDecoratableTransformer):
    def __init__(self):
        ...

    
    @m.call_if_inside(m.Assign(targets=[m.AssignTarget(m.Name("DATA"))]))
    def visit_Dict(self, node):
        print(node)

```

<h3> Repro Steps </h3>

 Inherit from the Visitor or Transform classes of libcst, then attempt to override functions within them.

https://github.com/microsoft/pylance-release/assets/25476118/bb54521c-4151-4e95-85f3-6a0e597fc599

<h3> Expected behavior </h3>

Code completion for overridden function names should be as fast as analyzing super.

https://github.com/microsoft/pylance-release/assets/25476118/2e9085ff-da14-4328-89de-1f893f5376f6

<h3> Actual behavior </h3>

After entering a function name, it takes about half a minute for the code completion suggestions for function names to appear.

<h3> Logs </h3>

[log.log](https://github.com/microsoft/pylance-release/files/13696216/log.log)

**Comment by Viicos** [comment]
Got also hit by this when using LibCST. Interesting that it goes way faster when using `super`. Have you tried using the `super` first? Maybe results are cached in some way.

**Comment by heejaechang** [comment]
it is probably due to us generating override method signature eagerly rather than lazily. let me take a look

**Comment by heejaechang** [comment]
even after making things lazy, it still can cost over 3+ seconds. but I think this particular case is a bit extreme since it has more than 800+ overloads that can be fuzzy matched by `visit` prefix. but I think it should be usable now

**Comment by Aobanana-chan** [comment]
Thank you for your prompt response and the effort you put into addressing the issue. I'm glad to hear that the fix is in progress and will be available in the next pre-release.
Even though a delay of over 3 seconds is not ideal, I think that it's tolerable. 

On a related note, I found that downgrading Pylance to a previous version has improved the working conditions. Specifically, the versions I tested are as follows:
2023.10.50 about 4 seconds
2023.9.40 about 5 seconds
2023.6.40 about 3 seconds
2023.5.40 about 2 seconds
2023.4.40 about 1 second
2023.3.40 less than 1 second
2022.10.40 less than 1 second, faster than 2023.3.40

I hope this information is helpful, and I look forward to the upcoming pre-release with the fix.

**Comment by heejaechang** [comment]
that's because in `old pylance`, we didn't attempt to generate `import statements` and etc when overridden method has type annotations. recent one does and that is what making it slower.

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2023.12.102, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202312102-4-January-2024-PreRelease)
